src/Native/System.Native/pal_errno.cpp

```diff
@@ -0,0 +1,249 @@
+        case Error::PAL_ENOTTY:           return ENOTTY;
+        case Error::PAL_ENXIO:            return ENXIO;
+        case Error::PAL_EOVERFLOW:        return EOVERFLOW;
+        case Error::PAL_EOWNERDEAD:       return EOWNERDEAD;
+        case Error::PAL_EPERM:            return EPERM;
+        case Error::PAL_EPIPE:            return EPIPE;
+        case Error::PAL_EPROTO:           return EPROTO;
+        case Error::PAL_EPROTONOSUPPORT:  return EPROTONOSUPPORT;
+        case Error::PAL_EPROTOTYPE:       return EPROTOTYPE;
+        case Error::PAL_ERANGE:           return ERANGE;
+        case Error::PAL_EROFS:            return EROFS;
+        case Error::PAL_ESPIPE:           return ESPIPE;
+        case Error::PAL_ESRCH:            return ESRCH;
+        case Error::PAL_ESTALE:           return ESTALE;
+        case Error::PAL_ETIME:            return ETIME;
+        case Error::PAL_ETIMEDOUT:        return ETIMEDOUT;
+        case Error::PAL_ETXTBSY:          return ETXTBSY;
+        case Error::PAL_EXDEV:            return EXDEV;
+        case Error::PAL_ENONSTANDARD:     break; // fall through to assert
+    }
+
+    // We should not use this function to round-trip platform -> pal
+    // -> platform. It's here only to synthesize a platform number
+    // from the fixed set above. Note that the assert is outside the
+    // switch rather than in a default case default block because not
+    // having a default will trigger a warning (as error) if there's
+    // an enum value we haven't handled. Should that trigger, make
+    // note that there is probably a corresponding missing case in the
+    // other direction above, but the compiler can't warn in that case
+    // because the platform values are not part of an enum.
+    assert(false);
+    return -1;
+}
+
+extern "C"
+const char* StrErrorR(int32_t platformErrno, char* buffer, int32_t bufferSize)
+{
+    assert(buffer != nullptr);
+    assert(bufferSize > 0);
+
+    // Note that we must use strerror_r because plain strerror is not
+    // thread-safe.
+    //
+    // However, there are two versions of strerror_r:
+    //    - GNU:   char* strerror_r(int, char*, size_t);
+    //    - POSIX: int   strerror_r(int, char*, size_t);
+    //
+    // The former may or may not use the supplied buffer, and returns
+    // the error message string.  The latter stores the error message
+    // string into the supplied buffer.
+
+#if HAVE_GNU_STRERROR_R
+    return strerror_r(platformErrno, buffer, bufferSize);
+#else
+    // We deliberately ignore the return value here. The only three
+    // valid return values are 0 for success, EINVAL for an unknown
+    // errno value, and ERANGE if there's not enough buffer space
+    // provided. For EINVAL, it'll still fill the buffer with a
+    // reasonable error string (e.g. "Unknown error: 0x123"), and for
+    // ERANGE, it'll fill the buffer with as much of the error as it
+    // can and null-terminate it.
```

Comment:
In the case of the XSI-compliant strerror_r, we could do better with regards to ERANGE, such as by returning NULL if we get ERANGE, at which point the managed side could allocate a larger buffer and call it again.